pkg/controller/external.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	kubeclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/terraform-provider-runtime/pkg/api"
+	"github.com/crossplane/terraform-provider-runtime/pkg/client"
+	"github.com/crossplane/terraform-provider-runtime/pkg/plugin"
+)
+
+const (
+	errNotMyType                  = "managed resource is not a MyType custom resource"
+	errProviderNotRetrieved       = "provider could not be retrieved"
+	errProviderSecretNil          = "cannot find Secret reference on Provider"
+	errProviderSecretNotRetrieved = "secret referred in provider could not be retrieved"
+
+	errNewClient = "cannot create new Service"
+)
+
+type External struct {
+	KubeClient kubeclient.Client
+	Invoker    *plugin.Invoker
+	Callbacks  managed.ExternalClientFns
+	logger     logging.Logger
+	provider   *client.Provider
+}
+
+func (c *External) Observe(ctx context.Context, res resource.Managed) (managed.ExternalObservation, error) {
+	c.entryLog(res, "Observe")
+	gvk := res.GetObjectKind().GroupVersionKind()
+	c.logger.Debug(fmt.Sprintf("terraform.External.Observe: %s", gvk.String()))
+	if c.Callbacks.ObserveFn != nil {
+		return c.Callbacks.Observe(ctx, res)
+	}
+
+	ares, err := api.Read(c.provider, c.Invoker, res)
```

Comment:
Should these `api` CRUD functions just be inlined into their `External` equivalents? Or are you thinking we'll make them injectable so we can inject other variants (e.g. for testing) at some point?